In [2]:
#Bayes Theorem i
p_coffee_drinker = .65
p_cancer = .005
p_coffee_drinker_given_cancer = .85
p_cancer_given_coffee_drinker = p_coffee_drinker_given_cancer * p_cancer / p_coffee_drinker
# prints 0.006538461538461539
print(p_cancer_given_coffee_drinker)


0.006538461538461539


In [4]:
#SciPy for the binomial distribution
from scipy.stats import binom
n = 10
p = 0.9
for k in range(n + 1):
    probability = binom.pmf(k, n, p)
    print("{0} - {1}".format(k, probability))
#

0 - 9.999999999999977e-11
1 - 8.999999999999976e-09
2 - 3.6449999999999933e-07
3 - 8.747999999999988e-06
4 - 0.00013778099999999974
5 - 0.0014880347999999982
6 - 0.011160260999999989
7 - 0.05739562799999997
8 - 0.1937102444999998
9 - 0.38742048899999976
10 - 0.34867844010000015


### Beta distribution
The beta distribution is a type of probability distribution, which means the area
under the entire curve is 1.0, or 100%. To find a probability, we need to find the area
within a range.

In [ ]:
#Beta distribution using SciPy
from scipy.stats import beta
a = 8
b = 2
p = beta.cdf(.90, a, b)

In [5]:
from scipy.stats import beta
a = 8
b = 2
p = 1.0 - beta.cdf(.90, a, b)
# 0.22515902199999993
print(p)

0.22515902199999982


So this means out of 8/10 successful engine tests, there is only a 22.5% chance the
underlying success rate is 90% or greater. But there is about a 77.5% chance it is
less than 90%.

How would we calculate an area in the middle? What if I want to
find the probability my underlying rate of success is between 80% and 90%

In [7]:
#Beta distribution middle area using SciPy
from scipy.stats import beta
a = 8
b = 2
p = beta.cdf(.90, a, b) - beta.cdf(.80, a, b)
# 0.33863336200000016
print(p)

0.3386333619999998


### Cumulative Density Function (CDF)

In [8]:
#The normal distribution CDF 
from scipy.stats import norm
mean = 64.43
std_dev = 2.99
x = norm.cdf(64.43, mean, std_dev)
print(x) # prints 0.5

0.5


In [9]:
#Getting a middle range probability using the CDF
from scipy.stats import norm
mean = 64.43
std_dev = 2.99
x = norm.cdf(66, mean, std_dev) - norm.cdf(62, mean, std_dev)
print(x) # prints 0.4920450147062894

0.4920450147062894


In [12]:
#Using the inverse CDF (called ppf()) in Python
from scipy.stats import norm
x = norm.ppf(.95, loc=64.43,scale=2.99)
print(x) # 69.3481123445849

69.3481123445849


In [ ]:
#Generating random numbers from a normal distribution
import random
from scipy.stats import norm
for i in range(0,1000):
    random_p = random.uniform(0.0, 1.0)
    random_weight = norm.ppf(random_p, loc=64.43, scale=2.99)
    #print(random_weight)



AttributeError: 'numpy.float64' object has no attribute 'plot'

In [16]:
#Turn Z-scores in x values and vice versa ¯
def z_score(x, mean, std):
    return (x - mean) / std

def z_to_x(z, mean, std):
    return (z * std) + mean
    mean = 140000
    std_dev = 3000
    x = 150000
# Convert to Z-score and then back to X
z = z_score(x, mean, std_dev)
back_to_x = z_to_x(z, mean, std_dev)
print("Z-Score: {}".format(z)) # Z-Score: 3.333
print("Back to X: {}".format(back_to_x)) # Back to X: 150000.0

Z-Score: 1.6448536269514709
Back to X: 69.3481123445849


## Inferential statistics

In [23]:
#Exploring the central limit theorem in Python
# Samples of the uniform distribution will average out to a normal distribution.
import random
import plotly.express as px
sample_size = 31
sample_count = 1000
# Central limit theorem, 1000 samples each with 31 random numbers between 0.0 and 1.0
x_values = [
    sum([random.uniform(0.0, 1.0) for _ in range(sample_size)]) / sample_size
    for _ in range(sample_count)
]


px.histogram(x=x_values, nbins=20, title="Central Limit Theorem Demo").show()

In [25]:
# Retrieving a critical Z-value
from scipy.stats import norm
def critical_z_value(p):
    norm_dist = norm(loc=0.0, scale=1.0)
    left_tail_area = (1.0 - p) / 2.0
    upper_area = 1.0 - ((1.0 - p) / 2.0)
    return norm_dist.ppf(left_tail_area), norm_dist.ppf(upper_area)

print(critical_z_value(p=.95))
# (-1.959963984540054, 1.959963984540054)

(np.float64(-1.959963984540054), np.float64(1.959963984540054))


In [27]:
#Calculating a confidence interval 
#sample of 31 objects has a mean of 64.408 and standard deviation of 2.05
from math import sqrt
from scipy.stats import norm
def critical_z_value(p):
    norm_dist = norm(loc=0.0, scale=1.0)
    left_tail_area = (1.0 - p) / 2.0
    upper_area = 1.0 - ((1.0 - p) / 2.0)
    return norm_dist.ppf(left_tail_area), norm_dist.ppf(upper_area)

def confidence_interval(p, sample_mean, sample_std, n):
# Sample size must be greater than 30
    lower, upper = critical_z_value(p)
    lower_ci = lower * (sample_std / sqrt(n))
    upper_ci = upper * (sample_std / sqrt(n))
    return sample_mean + lower_ci, sample_mean + upper_ci

print(confidence_interval(p=.95, sample_mean=64.408, sample_std=2.05,n=31))
# (63.68635915701992, 65.12964084298008)

(np.float64(63.68635915701992), np.float64(65.12964084298008))


In [28]:
#Calculating the probability of recovery between 15 and 21 days
from scipy.stats import norm
# Cold has 18 day mean recovery, 1.5 std dev
mean = 18
std_dev = 1.5
# 95% probability recovery time takes between 15 and 21 days.
x = norm.cdf(21, mean, std_dev) - norm.cdf(15, mean, std_dev)
print(x) # 0.9544997361036416

0.9544997361036416


#### One tailed test

In [ ]:
#getting x-value with 5% of area behind it
from scipy.stats import norm
# Cold has 18 day mean recovery, 1.5 std dev
mean = 18
std_dev = 1.5
# What x-value has 5% of area behind it?
x = norm.ppf(.05, mean, std_dev)
print(x) # 15.53271955957279

15.53271955957279


In [30]:
#Calculating the 1-tailed p-value
from scipy.stats import norm
# Cold has 18 day mean recovery, 1.5 std dev
mean = 18
std_dev = 1.5
# Probability of 16 or less days
p_value = norm.cdf(16, mean, std_dev)
print(p_value) # 0.09121121972586788

0.09121121972586788


#### Two-Tailed Test
To do a two-tailed test, we frame our null and alternative hypothesis in an
“equal” and “not equal” structure. In our drug test, we will say the null
hypothesis has a mean recovery time of 18 days. But our alternative hypothesis
is the mean recovery time is not 18 days thanks to the new drug.

In [31]:
from scipy.stats import norm
# Cold has 18 day mean recovery, 1.5 std dev
mean = 18
std_dev = 1.5
# What x-value has 2.5% of area behind it?
x1 = norm.ppf(.025, mean, std_dev)
# What x-value has 97.5% of area behind it
x2 = norm.ppf(.975, mean, std_dev)
print(x1) # 15.060054023189918
print(x2) # 20.93994597681008

15.060054023189918
20.93994597681008


In [32]:
#Calculating the 2-tailed p-value
from scipy.stats import norm
# Cold has 18 day mean recovery, 1.5 std dev
mean = 18
std_dev = 1.5
# Probability of 16 or less days
p1 = norm.cdf(16, mean, std_dev)
# Probability of 20 or more days
p2 = 1.0 - norm.cdf(20, mean, std_dev)
# P-value of both tails
p_value = p1 + p2
print(p_value) # 0.18242243945173575

0.18242243945173575


## T distribution

In [ ]:
from scipy.stats import t
# get critical value range for 95% confidence
# with a sample size of 25
n = 25
lower = t.ppf(.025, df=n-1)
upper = t.ppf(.975, df=n-1)
print(lower, upper)
#-2.063898561628021 2.0638985616280205

-2.063898561628021 2.0638985616280205


: 